### PR TITLE
Feature #36 - Complete email functionality

### PR DESCRIPTION
**Pull request by jost95**
Fixes: #36

The email functionality should now be complete. The system utilizes HTML
template with dynamic content for neatly structured emails. All
implementation details are hidden behind MailHandler.

**Usage example:**
```
MailHandler m = new MailHandler();
m.welcomeUser(new User(...));
```

### CODE REVIEW

File: src/main/java/se/lth/base/server/mail/CustomEmail.java [hunk 1]
```diff
@@ -2,64 +2,510 @@
...
+    }
+
+    private String parseResource(String template, Map<String, String> replacements) {
+        // Matches all instances of Mustache wraps
+        Matcher m = Pattern.compile("\\{\\{[A-z]\\w+}}").matcher(template);
+
+        while (m.find()) {
+            String mustacheValue = m.group().substring(2,m.group().length() - 2);
+            String replacement = replacements.get(mustacheValue);
+            System.out.println(replacement);
```
Comment (jost95) [inline src/main/java/se/lth/base/server/mail/CustomEmail.java]: Remove output

File: src/main/java/se/lth/base/server/mail/CustomEmail.java [hunk 2]
```diff
@@ -2,64 +2,510 @@
...
+
+
+        try (BufferedReader br = new BufferedReader(streamReader)) {
+            return br.lines().collect(Collectors.joining(System.lineSeparator()));
+        } catch (IOException e) {
+            LoggerFactory.getLogger(CustomEmail.class).error(e.getMessage(), e);
+        }
+
+        return "Error: could not read email template";
+    }
```
Comment (antondelorme) [inline src/main/java/se/lth/base/server/mail/CustomEmail.java]: `getResource` is only used in combination with `parseResource`, and vice versa. Combine the two.

Also, returning error as a string is bad practice; throw error instead. As I'm reading it now, we're still sending emails even if we can't read template (?).

File: src/main/java/se/lth/base/server/mail/CustomEmail.java [hunk 3]
```diff
@@ -2,64 +2,510 @@
...
+        for (DriveUser u : driveWrap.getUsers()) {
+            if (u.isDriver()) {
+                driver = u;
+            }
+
+            acceptedPassenger = u;
+        }
+
+        Timestamp departureTime = driveWrap.getDrive().getDepartureTime();
+        String formattedDepartureTime = new SimpleDateFormat("EEEEE MMMMM d").format(departureTime);
```
Comment (antondelorme) [inline src/main/java/se/lth/base/server/mail/CustomEmail.java]: Some `departureTime` are formatted to include time as well. Others just include date, like this one. If you only want to present date; use `formattedDepartureDate` for clarity.

File: src/main/java/se/lth/base/server/mail/CustomEmail.java [hunk 4]
```diff
@@ -2,64 +2,510 @@
...
-		StringBuilder sb = new StringBuilder();
-		
-		sb.append("<h1>Hi ").append(firstName).append("</h1>");
-		sb.append("<p>Welcome to CommuteCompanion!</p>");
-		
-		return sb.toString();
-	}
+    private final static String TITLE = "title", PREVIEW = "preview", INTRO = "intro",
+            BUTTON = "button", OUTRO = "outro", WEBSITE_LINK = "http://www.yourcommutecompanion.herokuapp.com",
+            EMAIL_TEMPLATE = "email-template.html";
```
Comment (antondelorme) [inline src/main/java/se/lth/base/server/mail/CustomEmail.java]: Should avoid single line declarations like this. It makes debugging more difiicult and worsens the readability.

File: src/main/java/se/lth/base/server/mail/CustomEmail.java [hunk 5]
```diff
@@ -2,64 +2,510 @@
...
+        return parseResource(getResource(EMAIL_TEMPLATE), content);
+    }
+
+    private String getBookingConfirmedSubject() {
+        return "Your booking request is confirmed";
+    }
+
+    private String getBookingConfirmedBody() {
+        Map<String, String> content = new HashMap<>();
+        DriveUser acceptedPassenger = null, driver = null;
```
Comment (antondelorme) [inline src/main/java/se/lth/base/server/mail/CustomEmail.java]: Separate

File: src/main/java/se/lth/base/server/mail/CustomEmail.java [hunk 6]
```diff
@@ -2,64 +2,510 @@
...
+        return parseResource(getResource(EMAIL_TEMPLATE), content);
+    }
+
+    private String getNewPassengerOnTripSubject() {
+        return "New passenger request for your drive to " + driveWrap.getDrive().getStop();
+    }
+
+    private String getNewPassengerOnTripBody() {
+        Map<String, String> content = new HashMap<>();
+        DriveUser lastPassenger = null, driver = null;
```
Comment (antondelorme) [inline src/main/java/se/lth/base/server/mail/CustomEmail.java]: Separate the declarations, for readability.

File: src/main/java/se/lth/base/server/mail/CustomEmail.java [hunk 7]
```diff
@@ -2,64 +2,510 @@
...
+        return parseResource(getResource(EMAIL_TEMPLATE), content);
+    }
+
+    private String getPassengerCancelledTripSubject() {
+        return "A passenger has cancelled their trip";
+    }
+
+    private String getPassengerCancelledTripBody() {
+        Map<String, String> content = new HashMap<>();
+        DriveUser lastPassenger = null, driver = null;
```
Comment (antondelorme) [inline src/main/java/se/lth/base/server/mail/CustomEmail.java]: Separate

File: src/main/java/se/lth/base/server/mail/CustomEmail.java [hunk 8]
```diff
@@ -2,64 +2,510 @@
...
+        return parseResource(getResource(EMAIL_TEMPLATE), content);
+    }
+
+    private String getDriverRemovedPassengerSubject() {
+        return "You have been removed from your trip";
+    }
+
+    private String getDriverRemovedPassengerBody() {
+        Map<String, String> content = new HashMap<>();
+        DriveUser removedPassenger = null, driver = null;
```
Comment (antondelorme) [inline src/main/java/se/lth/base/server/mail/CustomEmail.java]: Separate

File: src/main/java/se/lth/base/server/mail/CustomEmail.java [hunk 9]
```diff
@@ -2,64 +2,510 @@
...
+            if (u.isDriver()) {
+                driver = u;
+            }
+
+            lastPassenger = u;
+        }
+
+        // Get the user data
+        UserDataAccess userDao = new UserDataAccess(Config.instance().getDatabaseDriver());
+        assert driver != null;
```
Comment (antondelorme) [inline src/main/java/se/lth/base/server/mail/CustomEmail.java]: Bad practice to use `assert`. It is used by devs to find bugs during development. This code should be triggered by production code, i.e. user action. If for example driver somehow is null, we need to handle that exception (in this case, throw an exception from method).
There are multiple `assert`, won't comment them all but i think most (all?) can be replaced by
```java
if(var == null) {
   throw new IllegalArgumentException ("var is null");
}
```

File: src/main/resources/se/lth/base/server/mail/email-template.html [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+                    <!-- START MAIN CONTENT AREA -->
+                    <tr>
+                        <td class="wrapper">
+                            <table border="0" cellpadding="0" cellspacing="0">
+                                <tr>
+                                    <td>
+                                        {{intro}}
+                                        {{button}}
+                                        {{outro}}
+                                        <p>Best regards,<br>Jon at ComputeCompanion</p>
```
Comment (antondelorme) [inline src/main/resources/se/lth/base/server/mail/email-template.html]: Should all emails be from you?

Comment (jost95) [reply src/main/resources/se/lth/base/server/mail/email-template.html]: Or some imaginary name, I'll change it Alex

**Pull request merged**